content/kb/tutorials/databases/tidb.md

```diff
@@ -101,43 +106,47 @@ Copy the code below to your Streamlit app and run it. Make sure to adapt `query`
 ```python
 # streamlit_app.py
 
-import MySQLdb
 import streamlit as st
 
 # Initialize connection.
-# Uses st.cache_resource to only run once.
-@st.cache_resource
-def init_connection():
-    config = st.secrets["tidb"]
-    return MySQLdb.connect(
-        host=config["host"],
-        port=config["port"],
-        user=config["user"],
-        password=config["password"],
-        database=config["database"],
-        ssl_mode="VERIFY_IDENTITY",
-        ssl={"ca": config["ssl_ca"]}
-    )
-
-conn = init_connection()
+conn = st.experimental_connection('tidb', type='sql')
 
 # Perform query.
-# Uses st.cache_data to only rerun when the query changes or after 10 min.
-@st.cache_data(ttl=600)
-def run_query(query):
-    with conn.cursor() as cur:
-        cur.execute(query)
-        return cur.fetchall()
-
-rows = run_query("SELECT * from mytable;")
+df = conn.query('SELECT * from mytable;', ttl=600)
 
 # Print results.
-for row in rows:
-    st.write(f"{row[0]} has a :{row[1]}:")
+for row in df.itertuples():
+    st.write(f"{row.name} has a :{row.pet}:")
 ```
 
-See `st.cache_data` above? Without it, Streamlit would run the query every time the app reruns (e.g. on a widget interaction). With `st.cache_data`, it only runs when the query changes or after 10 minutes (that's what `ttl` is for). Watch out: If your database updates more frequently, you should adapt `ttl` or remove caching so viewers always see the latest data. Learn more in [Caching](/library/advanced-features/caching).
+See `st.experimental_connection` above? This handles secrets retrieval, setup, query caching and retries. By default, `query()` results are cached without expiring. In this case, we set `ttl=600` to ensure the query result is cached for no longer than 10 minutes. You can also set `ttl=0` to disable caching. Learn more in [Caching](/library/advanced-features/caching).
 
 If everything worked out (and you used the example table we created above), your app should look like this:
 
 ![Finished app screenshot](/images/databases/streamlit-app.png)
+
+## Connect with PyMySQL
+
+Other than [mysqlclient](https://github.com/PyMySQL/mysqlclient), [PyMySQL](https://github.com/PyMySQL/PyMySQL) is another popular MySQL Python client. To use PyMySQL, first you need to adapt your requirements file:
+
+```bash
+# requirements.txt
+PyMySQL==x.x.x
+SQLAlchemy==x.x.x
+```
+
+Then adapt your secrets file:
+
+```toml
+# .streamlit/secrets.toml
+
+[connections.tidb]
+dialect = "mysql"
+driver = "pymysql"
+host = "<TiDB_cluster_host>"
+port = 4000
+database = "pets"
+username = "<TiDB_cluster_user>"
+password = "<TiDB_cluster_password>"
+create_engine_kwargs = { connect_args = { ssl = { ca = "<path_to_CA_store>" }}}
```

Comment:
Oh this is interesting, I haven't tried creating the nested dict like this before in toml

@zhangyangyu did you confirm this works with the example app?

Reply:
Yes. I tried. Nested dict in toml could only be on one line, easy to be wrong :-).